src/ray/object_manager/connection_pool.cc

```diff
@@ -0,0 +1,165 @@
+  object_directory_ = object_directory;
+  connection_service_ = connection_service;
+  client_id_ = client_id;
+}
+
+void ConnectionPool::RegisterReceiver(ConnectionType type, const ClientID &client_id,
+                                      std::shared_ptr<ReceiverConnection> &conn) {
+  switch (type) {
+  case MESSAGE: {
+    Add(message_receive_connections_, client_id, conn);
+  } break;
+  case TRANSFER: {
+    Add(transfer_receive_connections_, client_id, conn);
+  } break;
+  }
+}
+
+void ConnectionPool::RemoveReceiver(ConnectionType type, const ClientID &client_id,
+                                    std::shared_ptr<ReceiverConnection> &conn) {
+  switch (type) {
+  case MESSAGE: {
+    Remove(message_receive_connections_, client_id, conn);
+  } break;
+  case TRANSFER: {
+    Remove(transfer_receive_connections_, client_id, conn);
+  } break;
+  }
+  // TODO(hme): appropriately dispose of client connection.
+}
+
+ray::Status ConnectionPool::GetSender(ConnectionType type, ClientID client_id,
+                                      SuccessCallback success_callback,
+                                      FailureCallback failure_callback) {
+  SenderMapType &avail_conn_map = (type == MESSAGE)
+                                      ? available_message_send_connections_
+                                      : available_transfer_send_connections_;
+  if (Count(avail_conn_map, client_id) > 0) {
+    success_callback(Borrow(avail_conn_map, client_id));
+    return Status::OK();
+  } else {
+    return CreateConnection1(
+        type, client_id,
+        [this, type, client_id, success_callback](SenderConnection::pointer conn) {
+          // add it to the connection map to maintain ownership.
+          SenderMapType &conn_map =
+              (type == MESSAGE) ? message_send_connections_ : transfer_send_connections_;
+          Add(conn_map, client_id, conn);
+          // add it to the available connections
+          SenderMapType &avail_conn_map = (type == MESSAGE)
+                                              ? available_message_send_connections_
+                                              : available_transfer_send_connections_;
+          Add(avail_conn_map, client_id, conn);
+          // "borrow" the connection.
+          success_callback(Borrow(avail_conn_map, client_id));
+        },
+        failure_callback);
+  }
+}
+
+ray::Status ConnectionPool::ReleaseSender(ConnectionType type,
+                                          SenderConnection::pointer conn) {
+  SenderMapType &conn_map = (type == MESSAGE) ? available_message_send_connections_
+                                              : available_transfer_send_connections_;
+  Return(conn_map, conn->GetClientID(), conn);
+  return ray::Status::OK();
+}
+
+void ConnectionPool::Add(ReceiverMapType &conn_map, const ClientID &client_id,
+                         std::shared_ptr<ReceiverConnection> conn) {
+  if (conn_map.count(client_id) == 0) {
```

Comment:
This is not necessary. `conn_map[client_id]` will perform the empty insert if the key does not exist.